Importing the necessary modules for the exercise.

In [5]:
import requests
import pandas as pd
from bs4 import BeautifulSoup
from IPython.core.display import HTML

Make a request using `requests.get()` method to the Wikipedia page. Store the text from the page in a local variable

In [2]:
req = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
page = req.text

Use BeautifulSoup's `html.parser` to parse the page content

In [3]:
soup = BeautifulSoup(page, 'html.parser')

Find all the tables in the page. There might be many tables and we are interested in only one of the tables.

In [4]:
can_table = soup.find_all("table")
print(len(can_table))

3


In [6]:
HTML(str(can_table[0]))

We see that there are three tables in the page and the first of those seems to be the one we want.

Let us look at the table we extracted to see what other HTML elements need to be filtered out.

In [7]:
can_table[0]

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td

We see that the columns names are enclosed within `<th>` and the values of the different rows and columns (the names of the Boroughs, Neighborhoods and the postal codes) are enclosed withint `<td>` tags.

Let us filter these one by one.

In [9]:
columns = [c for c in can_table[0].find_all('th')]
columns

[<th>Postal Code
 </th>,
 <th>Borough
 </th>,
 <th>Neighbourhood
 </th>]

Let us write a lambda function to replace new line characters with just a space

In [10]:
rm_newline = lambda s: s.replace("\n", "")

In [11]:
columns = [rm_newline(c.get_text()) for c in columns]
columns

['Postal Code', 'Borough', 'Neighbourhood']

Let us do a similar thing for filtering out rows (which are in `<td>`)

In [12]:
rows = [r for r in can_table[0].find_all('td')]
rows

[<td>M1A
 </td>,
 <td>Not assigned
 </td>,
 <td>Not assigned
 </td>,
 <td>M2A
 </td>,
 <td>Not assigned
 </td>,
 <td>Not assigned
 </td>,
 <td>M3A
 </td>,
 <td>North York
 </td>,
 <td>Parkwoods
 </td>,
 <td>M4A
 </td>,
 <td>North York
 </td>,
 <td>Victoria Village
 </td>,
 <td>M5A
 </td>,
 <td>Downtown Toronto
 </td>,
 <td>Regent Park, Harbourfront
 </td>,
 <td>M6A
 </td>,
 <td>North York
 </td>,
 <td>Lawrence Manor, Lawrence Heights
 </td>,
 <td>M7A
 </td>,
 <td>Downtown Toronto
 </td>,
 <td>Queen's Park, Ontario Provincial Government
 </td>,
 <td>M8A
 </td>,
 <td>Not assigned
 </td>,
 <td>Not assigned
 </td>,
 <td>M9A
 </td>,
 <td>Etobicoke
 </td>,
 <td>Islington Avenue, Humber Valley Village
 </td>,
 <td>M1B
 </td>,
 <td>Scarborough
 </td>,
 <td>Malvern, Rouge
 </td>,
 <td>M2B
 </td>,
 <td>Not assigned
 </td>,
 <td>Not assigned
 </td>,
 <td>M3B
 </td>,
 <td>North York
 </td>,
 <td>Don Mills
 </td>,
 <td>M4B
 </td>,
 <td>East York
 </td>,
 <td>Parkview Hill, Woodbine Gardens
 </td>,


Let us replace newline characters with spaces

In [13]:
rows = [rm_newline(r.get_text()) for r in rows]
rows

['M1A',
 'Not assigned',
 'Not assigned',
 'M2A',
 'Not assigned',
 'Not assigned',
 'M3A',
 'North York',
 'Parkwoods',
 'M4A',
 'North York',
 'Victoria Village',
 'M5A',
 'Downtown Toronto',
 'Regent Park, Harbourfront',
 'M6A',
 'North York',
 'Lawrence Manor, Lawrence Heights',
 'M7A',
 'Downtown Toronto',
 "Queen's Park, Ontario Provincial Government",
 'M8A',
 'Not assigned',
 'Not assigned',
 'M9A',
 'Etobicoke',
 'Islington Avenue, Humber Valley Village',
 'M1B',
 'Scarborough',
 'Malvern, Rouge',
 'M2B',
 'Not assigned',
 'Not assigned',
 'M3B',
 'North York',
 'Don Mills',
 'M4B',
 'East York',
 'Parkview Hill, Woodbine Gardens',
 'M5B',
 'Downtown Toronto',
 'Garden District, Ryerson',
 'M6B',
 'North York',
 'Glencairn',
 'M7B',
 'Not assigned',
 'Not assigned',
 'M8B',
 'Not assigned',
 'Not assigned',
 'M9B',
 'Etobicoke',
 'West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale',
 'M1C',
 'Scarborough',
 'Rouge Hill, Port Union, Highland Creek',
 'M2C',


Now, we have a flattened out list of all the elements (content) in the table and we want to organize them under the columns. The elements (content) in each row is organized within these three columns. So, we need to group them in groups of three (three = number of columns).

In [14]:
rows = list(zip(*[rows[i::3] for i in range(len(columns))]))
rows

[('M1A', 'Not assigned', 'Not assigned'),
 ('M2A', 'Not assigned', 'Not assigned'),
 ('M3A', 'North York', 'Parkwoods'),
 ('M4A', 'North York', 'Victoria Village'),
 ('M5A', 'Downtown Toronto', 'Regent Park, Harbourfront'),
 ('M6A', 'North York', 'Lawrence Manor, Lawrence Heights'),
 ('M7A', 'Downtown Toronto', "Queen's Park, Ontario Provincial Government"),
 ('M8A', 'Not assigned', 'Not assigned'),
 ('M9A', 'Etobicoke', 'Islington Avenue, Humber Valley Village'),
 ('M1B', 'Scarborough', 'Malvern, Rouge'),
 ('M2B', 'Not assigned', 'Not assigned'),
 ('M3B', 'North York', 'Don Mills'),
 ('M4B', 'East York', 'Parkview Hill, Woodbine Gardens'),
 ('M5B', 'Downtown Toronto', 'Garden District, Ryerson'),
 ('M6B', 'North York', 'Glencairn'),
 ('M7B', 'Not assigned', 'Not assigned'),
 ('M8B', 'Not assigned', 'Not assigned'),
 ('M9B',
  'Etobicoke',
  'West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale'),
 ('M1C', 'Scarborough', 'Rouge Hill, Port Union, Highland Creek'),
 ('M

We have all the ingredients in the form of python lists. We can use pandas to convert these lists into a dataframe

In [17]:
df = pd.DataFrame(rows, columns=columns)
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Now, we have the dataframe to play with. We can use pandas to filter out further. First, let us drop all the rows in which Borough is "Not assigned"

In [22]:
not_assigned_indexes = df[df['Borough']=='Not assigned'].index
df.drop(not_assigned_indexes, inplace=True)
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Now, let us merge the rows that have the same postal code (rows grouped by postal code) into one row, collecting the neighborhoods belonging to the postal code into a list (achieved by a join using ',')

In [23]:
df = df.groupby(df['Postal Code']).agg({'Borough': 'first', 'Neighbourhood' : ', '.join}).reset_index()

In [24]:
df.head()

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

How many rows are there?

In [25]:
df.shape

(103, 3)

So, our dataset has 103 rows organized under 3 columns